[Avatar] Implement Cait Sith

MowFord opened the pull request:
**_I affirm:_**
- [x] I understand that if I do not agree to the following points by completing the checkboxes my PR will be ignored.
- [x] I understand I should leave resolving conversations to the LandSandBoat team so that reviewers won't miss what was said.
- [x] I have read and understood the [Contributing Guide](https://github.com/LandSandBoat/server/blob/base/CONTRIBUTING.md) and the [Code of Conduct](https://github.com/LandSandBoat/server/blob/base/CODE_OF_CONDUCT.md).
- [x] I have _**tested my code and the things my code has changed**_ since the last commit in the PR and will test after any later commits.

## What does this pull request do?

Implements the Cait Sith smn avatar!

~~I'm starting this off as a draft to get feedback on the core changes I've had to make for Level ? Holy, as well as the path I've taken for the pet skills. Specifically, the buff spells don't report in the chat log properly~~

~~I'm running into a weird situation with using `xi.msg.basic.SKILL_GAIN_EFFECT` when these are mobskills. It makes sense, as it's likely reporting the player ability ID as the mob skill ID. If i set the `pet_skill_finish_id` as player JA finish, then the chat log reports properly but the animation is wrong <.<~~

~~I've got the buff abilities to work via `xi.msg.basic.IS_STATUS`, but it isn't the cleanest.~~

~~Note that I've not adjusted the meat of the offensive abilities to actually function, I'm merely making sure I'm on the right track here. It will technically work this way, but I want to make sure it's the way LSB would prefer.~~

Level ? Holy working via clay's adjustments:
This was a level 5, animation was proper and it properly had no effect on the mob that wasn't divisible by 5
![image](https://github.com/LandSandBoat/server/assets/131182600/edcab9da-f2e1-4677-b15f-5092d887b97d)

Onto a structured list of the changes:
- Cait sith has proper name prefix and named properly to be "Cait Sith" instead of "The CaitSith"
  - Done via proper name in `pet_list` and `name_prefix` in `mob_pools` set to 32
- abilities enabled in `abilities.sql` and set to have proper MP cost (and other metadata)
  -  Also fixed a bug with one of fenrir's abilities being grouped in when Cait Sith was the player's pet
- lua files and entries in the new `pet_skills.sql` table created for all blood pacts:
  - Regal Slash (BP:Rage): 3-hit physical
  - Level ? Holy (BP:Rage): aoe magical
    - Rolls a die and does dmg proportional to roll
    - Only does damage if the target's level is divisible by the roll
  - Mewing Lullaby (BP:Ward): AoE lullaby that resets TP on non-NM mobs
  - Eerie Eye (BP:Ward): conal silence/amnesia with appropriate elemental resist check for amnesia, but retail does light check for silence 
  - Reraise II (BP:Ward): single-target 60-minute reraise II buff for any party member
  - Raise II (BP:Ward): single-target raise II for any party member
  - Altana's Favor (BP:Ward): 2-hour ability gives arise to all party members in range (raise 5 and reraise III with infinite duration)
    - couldn't get targetfind to work properly between dead and alive players, so I loop over all party members and raise them manually

## Steps to test these changes

This PR does not include the changes from [clay's PR](https://github.com/LandSandBoat/server/pull/4627/files), but was tested using the changes

Testing of Level ? Holy is best done on mobs that range around lvl 60 since that level is divisible by all numbers 1-6.

The others are pretty straightforward I think. Giving your pet `!immortal` makes things easier to test the BP:Rages as well.

MowFord commented:
The crux of my issue with level ? holy is that there seems to be a fundamental flaw in how pet abilities process. The onabilitycheck is executed on the owner's ability, so you can't edit the animation ID of the petskill. (i.e. like I [do in this mobskill](https://gitlab.com/ffxiwings/wings/-/blob/master/scripts/globals/mobskills/self-destruct_cluster.lua#L46))

WinterSolstice8 commented:
I'm busy with other stuff, I wrote the petskill system and there _should_ be a way to iron out the kinks, i'll try to look into this at some point on the weekend. Thanks for your efforts

MowFord commented:
> I'm busy with other stuff, I wrote the petskill system and there _should_ be a way to iron out the kinks, i'll try to look into this at some point on the weekend. Thanks for your efforts

Sounds good. I'm going to work on cleaning up the lua I just wanted to get something in front of you so I'm not completely in the dark.

Also just to be clear: this code works perfectly but
- I'm not a fan of hard coding the animation changes in cpp
- the messages don't work perfectly as it seems the ability id is referencing mob skills instead of pet skills (i.e. if we set reraise_ii to use message id 317, it says cait sith used berserk, because berserk is mob skill id 526

WinterSolstice8 commented:
Battlemod doesn't seem to like Cait Sith (LSB image)
![image](https://github.com/LandSandBoat/server/assets/60417494/9255f7ce-436b-4257-aab2-afd6c2e7d590)
testing retail now...

WinterSolstice8 commented:
Cait Sith's name is wrong (missing the space) and the pool/group has "The" flagged, retail msg:

![image](https://github.com/LandSandBoat/server/assets/60417494/f6548fb3-2986-4dc3-9eee-c4bef3562934)

Level ? Holy needs to have the primary message set to `skill:setMsg(xi.msg.basic.JA_NO_EFFECT_2)` if it misses per a retail cap

![image](https://github.com/LandSandBoat/server/assets/60417494/e6cadd16-9db3-4ef7-b37d-37ae5aa0ba26)

WinterSolstice8 commented:
Reraise II message is `JA_GAIN_EFFECT_2       = 316, -- <user> uses <ability>. <target> gains the effect of <effect>.`
(previously undocumented in msg.lua)

WinterSolstice8 commented:
I see how to fix the problem with level ? holy, because the animation (presumably) stores the "roll" for what the level ? holy will do damage to, and rerolling it is not retail behavior.

I think I just need to add an accessor to get the action's first target and from that, you'll be able to get the action's first animation and/or set it appropriately. I'll hopefully get a PR tomorrow for the change.

WinterSolstice8 commented:
Level ? holy script works with some backend changes like this:

```
abilityObject.onPetAbility = function(target, pet, skill, master, action)
    local holyRollOneAnimID = 164
    local primaryTargetID = action:getPrimaryTargetID()

    -- If primary target, roll for power by setting random animation.
    if primaryTargetID == target:getID() then
        action:setAnimation(primaryTargetID, holyRollOneAnimID + math.random(0, 5))
    else
        action:setAnimation(target:getID(), action:getAnimation(primaryTargetID))
    end

    local power = action:getAnimation(target:getID()) - holyRollOneAnimID + 1
    local dMND = math.floor(pet:getStat(xi.mod.MND) - target:getStat(xi.mod.MND))
    local ele = xi.damageType.LIGHT
    local dmg = 0

    local dmgmod = 1
    local basedmg = pet:getMainLvl() * power + (dMND * 1.5)
```

MowFord commented:
Apologies if it fails sanity checks, have to step away before doing that step, but bps are tested to work

MowFord commented:
Unfortunately the issue i found with `altana's favor` is also happening for `raise ii`. I thought i had confirmed it works but I cannot get the ability to consider a dead target valid for the ability :(

I diagnosed a good bit and best I can tell the issue is related to `CCharEntity::ValidTarget`. I set the flag 32 for `target_player_dead` but it just doesn't even call the onAbilityCheck function

MowFord commented:
If it's simpler to get things moving we could disable Raise II in `abilities.sql` to get Cait Sith implemented the rest of the way for now. I know you'd rather not have half-implemented thing, but Cait Sith being unimplemented at all might be a special case?

WinterSolstice8 commented:
might be able to look at it later tonight, stay tuned

WinterSolstice8 commented:
![image](https://github.com/LandSandBoat/server/assets/60417494/5516bca9-ef64-480e-86d6-e24cc0c49a4a)
soon

WinterSolstice8 commented:
![image](https://github.com/LandSandBoat/server/assets/60417494/dde4d5ea-0df0-4be6-8cde-457e2a204145)

WinterSolstice8 commented:
Is there anything else you're missing? Work has been a bit busy.

MowFord commented:
> Is there anything else you're missing? Work has been a bit busy.

That was everything needed for this PR. I appreciate you. I'll clean up my PR to utilize your changes later this week

WinterSolstice8 commented on src/map/utils/charutils.cpp:
```diff
@@ -2915,7 +2915,7 @@ namespace charutils
                     else if (PetID == PETID_CAIT_SITH)
                     {
-                        if (PAbility->getID() > ABILITY_SOOTHING_RUBY && PAbility->getID() <= ABILITY_MOONLIT_CHARGE)
+                        if (PAbility->getID() > ABILITY_SOOTHING_RUBY && PAbility->getID() < ABILITY_MOONLIT_CHARGE)
```
Why was this changed? Was it necessary?

MowFord replied:
yes, when cait sith was out, moonlit charge was in the list of usable blood pacts

Xaver-DaRed commented on scripts/actions/abilities/pets/mewing_lullaby.lua:
```diff
@@ -0,0 +1,43 @@
+end
+
+abilityObject.onPetAbility = function(target, pet, skill, master, action)
+    local typeEffect = xi.effect.LULLABY
```
I know it's called lullaby. However, It was proven that lullaby BRD spells actually aply sleep effect, not lullaby effect, as weird as it sounds.

My point? Either we follow suit OR we confirm in retail the actual effect

MowFord replied:
I'm not 100% knowledgeable on the specifics, and we could apply sleep effect instead of lullaby, but doesn't the chat log explicitly say `lullaby` instead of `sleep`?

TeoTwawki replied:
We can easily use one in the addStatus and the other in the message return, that's not a hurdle. But if you think it actually uses the other status, that's easy to check with a packet cap.

MowFord replied:
Xaver DM'd me about it and that is exactly what we updated it to.

Xaver-DaRed commented on scripts/actions/spells/summoning/cait_sith.lua:
```diff
@@ -0,0 +1,32 @@
+-----------------------------------
+-- Spell: Diabolos
+-- Summons Diabolos to fight by your side
```
Nitpick: Wrong spell comment and description

MowFord replied:
Oops, good catch.

Guess you can see what I copied from :)

WinterSolstice8 commented on scripts/actions/abilities/pets/regal_scratch.lua:
```diff
@@ -0,0 +1,30 @@
+-----------------------------------
+-- Regal Scratch
+-- M=3,1
```
what's this `M=3,1`? doesn't really bother me that it's in there i guess

MowFord replied:
ha good question, was a direct copy from what we had on wings... i honestly have no idea

MowFord committed:
Cait Sith Avatar:

- Cait sith has proper name prefix and named properly to be "Cait Sith" instead of "The CaitSith"
- BPs Implemented
  - Regal Slash (BP:Rage): 3-hit physical
  - Level ? Holy (BP:Rage): aoe magical
    - Rolls a die and does dmg proportional to roll
    - Only does damage if the target's level is divisible by the roll
  - Mewing Lullaby (BP:Ward):
  AoE lullaby that resets TP
  - Eerie Eye (BP:Ward):
  conal silence/amnesia with appropriate elemental resist check for amnesia, but retail does light check for silence
  - Reraise II (BP:Ward):
  single-target 60-minute reraise II buff for any party member
  - Raise II (BP:Ward):
  single-target raise II for any party member
  - Altana's Favor (BP:Ward):
  2-hour ability gives arise to all party members in range
  (Arise and reraise III with infinite duration)
